### PR TITLE
Add support for mamba

### PR DESCRIPTION
**Pull request by marcelotrevisani**

### CODE REVIEW

**Comment by prusse-martin** [comment]
Nice.
Thank you for this =)

But I am curious. Do you know how mamba deals with the `truncate_history_file` workaround? I ask because in recent conda versions that cause the environment to not be pruned at all.

File: conda_devenv/__init__.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
-from ._version import version
-
-__version__ = version
+try:
+    from ._version import version as __version__
+except ModuleNotFoundError:
+    __version__ = "0.0.0"
```
Comment (prusse-martin) [inline conda_devenv/__init__.py:4]: Just a question.
Did you stumble into some problem with this?
Like testing the changes (obviously, in an unreleased) version?

Comment (marcelotrevisani) [reply conda_devenv/__init__.py:4]: Yeah, I was having some problems to run all the tests, specially those which test the ``conda-devenv`` version 
I can change if you want to, I just thought that it would be nice to able to run all tests without a release in place

**Comment by marcelotrevisani** [comment]
> Nice.
> Thank you for this =)
> 
> But I am curious. Do you know how mamba deals with the `truncate_history_file` workaround? I ask because in recent conda versions that cause the environment to not be pruned at all.

I am not sure, let me ping @wolfv , he may know better about it

File: conda_devenv/devenv.py [hunk 2]
```diff
@@ -387,11 +385,21 @@ def truncate_history_file(env_directory):
...
+    try:
+        import mamba
+    except ModuleNotFoundError:
+        return False
+    return True
+
+
 def __call_conda_env_update(args, output_filename):
-    import sys
+    env_manager = "conda"
```
Comment (nicoddemus) [inline conda_devenv/devenv.py:397]: Can we instead use `sys.arv[0]` to obtain which program is running? This would make this generic (without knowing about `mamba` at all), and simplify the code a lot.

If that works, we can even discard the `disable_mamba` option.

Comment (marcelotrevisani) [reply conda_devenv/devenv.py:397]: Sure, I am going to change it

File: conda_devenv/devenv.py [hunk 3]
```diff
@@ -419,7 +409,7 @@ def __call_conda_env_update(args, output_filename):
 
     old_argv = sys.argv[:]
     try:
-        sys.argv = command[1:] if env_manager == "conda" else command
+        sys.argv = command if env_manager == "mamba" else command[1:]
```
Comment (nicoddemus) [inline conda_devenv/devenv.py:412]: Weird, why is this needed?

Comment (marcelotrevisani) [reply conda_devenv/devenv.py:412]: Because ``mamba`` expects the first argument to be ``mamba``
:grimacing:

Comment (marcelotrevisani) [reply conda_devenv/devenv.py:412]: https://github.com/TheSnakePit/mamba/blob/77c0e3cf6284aa18be0d2bfbe5ef7c2de7b642d8/mamba/mamba.py#L761

Comment (marcelotrevisani) [reply conda_devenv/devenv.py:412]: actually, it can be anything, so I just let the full command

File: conda_devenv/devenv.py (hunk 2 above)
Comment (prusse-martin) [reply conda_devenv/devenv.py]: Sory... I don't think conda_devenv receive the info about the calling program from argv.

Comment (marcelotrevisani) [reply conda_devenv/devenv.py]: gonna test it later, but I am not so sure either :grimacing:

Comment (prusse-martin) [reply conda_devenv/devenv.py]:
```
W:\
λ cat \Miniconda\Lib\site-packages\conda_devenv\devenv.py | grep -C5 "============================="


def main(args=None):
    if args is None:
        args = sys.argv[1:]
    print('=============================')
    print(sys.argv)
    return 0
    parser = argparse.ArgumentParser(
        description="Work with multiple conda-environment-like yaml files in dev mode."
    )

W:\
λ conda devenv --version
=============================
['W:\\Miniconda\\Scripts\\conda-devenv', '--version']
```

Comment (marcelotrevisani) [reply conda_devenv/devenv.py]: Thanks @prusse-martin !
I will revert the modifications then!
I will do it later, because I am a bit busy right now :grimacing:

File: conda_devenv/__init__.py (hunk 1 above)
Comment (prusse-martin) [reply conda_devenv/__init__.py:4]: I agree with you about being able to run the tests.

**Comment by wolfv** [comment]
@prusse-martin

**Comment by wolfv** [comment]
@prusse-martin 

> But I am curious. Do you know how mamba deals with the truncate_history_file workaround? I ask because in recent conda versions that cause the environment to not be pruned at all.

Can you describe what you mean? How do you use conda? Mamba might currently also not correctly prune, but if you set update a complete environment, why not completely erase the previous one?

**Comment by nicoddemus** [comment]
> why not completely erase the previous one?

This is quick on Linux, but unfortunately on Windows it takes a few minutes to remove + create an environment for scratch.

**Comment by prusse-martin** [comment]
> Can you describe what you mean?

when updating an environment with a `environment.yml` using the `--prune` flag the resulting environment should have the same packages (and only them), not leaving formerly installed packages not required by the declared packages in `environment.yml`.

> How do you use conda?

With conda-devenv the resulting conda command is `conda env update --prune` always, even for creating the environment. In our in-house tools we use some other conda commands/API for things like caching artifacts.

> Mamba might currently also not correctly prune ...

Maybe this is some contribution we can make at some point =)

> ... but if you set update a complete environment, why not completely erase the previous one?

As @nicoddemus told on windows it is a lot faster to just update some small packages in a large environment then recreate it from scratch. And think on windows unlinking and link some packages with a lot of files (we are talking about things like boost and qt).

**Comment by prusse-martin** [comment]
Just some history.

If I remember right it was @nicoddemus that make the first implementation of prune in conda.
At some point the prune behavior was broken but some one devised a [workaround (truncating the history file)](https://github.com/conda/conda/issues/7279#issuecomment-389359679) (we add that workaround in conda_devenv).
On more recent versions prune will not remove packages and truncating the history does not work any more.

**Comment by phausamann** [comment]
Hi all, just wondering if this is ready to be merged?

Or is the pruning issue/truncating history file something that needs to be addressed? In that case, is there anything I can do to help move this along?

In any case, I'm excited to see conda devenv work with mamba!

**Comment by nicoddemus** [comment]
Hey @phausamann thanks for the ping!

@prusse-martin do you recall the details?

**Comment by prusse-martin** [comment]
We didn't check the `--prune` flag on mamba.
Without that some changes (removal of dependencies) will not be reflected when updating the conda environment.

I just did run some local tests and running `mamba env update --prune --file environment.yml` did result in the same env as conda.

@marcelotrevisani
Do you have any idea if mamba should impact `mamba env ...` or it just forward the call to vanilla conda? I used a small test and didn't time runs.

The main problem appears to be  with recent versions of conda where `--prune` is not longer working as we expected.
When just installing mamba my conda get upgraded to a recent version but "pinning" my conda it can install mamba, on both occasions I got mamba 0.1.2 (not the latest).

Any way, with prune working or not I think it is important to note it only affects updates, not when creating a new environment from scratch.

**Comment by wolfv** [comment]
@prusse-martin I don't think `--prune` for environment files makes a lot of sense. just recreate environment from scratch when updating (and rm -rf the old one).

At least that's how I'd implement a `--prune` flag in mamba (solve fresh from the environment.yml file).

**Comment by prusse-martin** [comment]
On windows recreating the env from scratch can take much more time than `update --prune` it there are packages with thousands of files.

**Comment by majidaldo** [comment]
> @prusse-martin I don't think `--prune` for environment files makes a lot of sense. just recreate environment from scratch when updating (and rm -rf the old one).
> 
> At least that's how I'd implement a `--prune` flag in mamba (solve fresh from the environment.yml file).

i think that's the `--force` option.

**Comment by prusse-martin** [comment]
In the conda `--force` option cause the environment to be removed and recreated, as far as I know the `--prune` option should result in the same resulting "environment state" without having to pay for unlinking then linking all files.

Think of an environment with over 60k files, on windows, and having to perform few updates to the env but having to pay a full unlink/link due `--force`

**Comment by rchomjak** [comment]
@prusse-martin Can you merge this PR?

**Comment by prusse-martin** [comment]
I did just test conda 4.10.3 and mamba 0.17.0, and I got the same behavior for prune (it was ignored, that is not very nice). conda-devenv expect `prune` to behave as listed in the help.

```
λ conda env update --help
usage: conda-env-script.py update [-h] [-n ENVIRONMENT | -p PATH] [-f FILE]
                                  [--prune] [--json] [-v] [-q]
                                  [remote_definition]

Update the current environment based on environment file
...
  --prune               remove installed packages not defined in
                        environment.yml
```

@nicoddemus, given recent versions of conda and mamba ignore the flag conda-devenv uses I will be in favor of merging it.

**Comment by nicoddemus** [comment]
If it is good for @prusse-martin, it is good for me then. 👍 

Please go ahead @prusse-martin.

**Comment by rchomjak** [comment]
@prusse-martin  and @nicoddemus thank you very much.

**Comment by joannasendorek** [comment]
Hi! Great changes in this PR and I am really interested in them for my usage. Could we please merge this PR?
